File: src/main/java/org/cip4/xjdf/json/openapi/LocalType.java
package org.cip4.xjdf.json.openapi;

import lombok.AllArgsConstructor;
import org.cip4.xjdf.json.openapi.model.Model;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;

@AllArgsConstructor
public class LocalType extends AttributeType {
    private Node node;
    private Context context;

    public Node getSimpleType() {
        try {
            return (Node) context.getXPath().evaluate("xs:simpleType", node, XPathConstants.NODE);
        } catch (Exception e) {
            throw new RuntimeException("Error evaluating XPath expression", e);
        }
    }

    @Override
    public Model getModel() {
        return SimpleType.create(getSimpleType(), context).getModel();
    }

    public static class Factory implements AttributeType.Factory<LocalType> {
        @Override
        public boolean supports(Node candidate, XPath xPath) {
            try {
                NodeList nodeList = (NodeList) xPath.evaluate("xs:simpleType", candidate, XPathConstants.NODESET);
                return nodeList.getLength() > 0;
            } catch (Exception e) {
                throw new RuntimeException("Error evaluating XPath expression", e);
            }
        }

        @Override
        public LocalType create(Node node, Context context) {
            return new LocalType(node, context);
        }
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/LocalType.kt
package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.Schema
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants

class LocalType(
    private val node: Node,
    private val context: Context
) : AttributeType() {

    val simpleType: Node
        get() = context.xPath.evaluate(
            "xs:simpleType",
            node,
            XPathConstants.NODE
        ) as Node

    companion object Factory : AttributeType.Factory<LocalType> {
        override fun supports(candidate: Node, xPath: XPath): Boolean =
            (xPath.evaluate(
                "xs:simpleType",
                candidate,
                XPathConstants.NODESET
            ) as NodeList).length > 0

        override fun create(node: Node, context: Context) = LocalType(node, context)
    }

    override fun getModel(): Schema {
        return SimpleType.Factory.create(simpleType, context).getModel()
    }
}